cleo_plugins/Audio/CAudioStream.h

```diff
@@ -5,7 +5,14 @@
 
 namespace CLEO
 {
-    class CAudioStream
+    // in CLEO4 CAudioStream was extending VInjectible
```

Comment:
No, it wasn't

https://github.com/cleolibrary/CLEO4/blob/master/source/CSoundSystem.h#L57

Reply:
Hm, right. 
Then why in CLEO4 offset was 8 and now without adjustments it will be just 4.

Reply:
Where did you find offset 8? 

```
.text:100080C0 ; __stdcall CLEO_GetInternalAudioStream(x, x)
.text:100080C0                 public _CLEO_GetInternalAudioStream@8
.text:100080C0 _CLEO_GetInternalAudioStream@8 proc near
.text:100080C0                                         ; DATA XREF: .rdata:off_1004A6B8↓o
.text:100080C0
.text:100080C0 arg_4           = dword ptr  0Ch
.text:100080C0
.text:100080C0                 push    ebp
.text:100080C1                 mov     ebp, esp
.text:100080C3                 mov     eax, [ebp+arg_4]
.text:100080C6                 mov     eax, [eax+4] <------------------------ +4
.text:100080C9                 pop     ebp
.text:100080CA                 retn    8
.text:100080CA _CLEO_GetInternalAudioStream@8 endp
```

Reply:
Ok I had to write my own test script:
```
{$CLEO .cs}
nop

int lib 
if
    not lib = load_dynamic_library {fileName} "bass.dll"
then
    terminate_this_script
end

int proc
if
    not proc = get_dynamic_library_procedure {procName} "BASS_ChannelGetLevel" {library} lib
then
    terminate_this_script
end

int stream = load_audio_stream {audioFileName} "cleo\cleo_tests\Audio\Ding.mp3"
//int stream = load_3d_audio_stream {audioFileName} "cleo\cleo_tests\Audio\Ding.mp3"

set_audio_stream_state stream {action} AudioStreamAction.Play
set_audio_stream_looped stream {state} true


while true
    wait 0
    
    int bassStream = stream + 4
    bassStream = read_memory {address} bassStream {size} 4 {vp} false
    
    int levels = call_function_return {address} proc {numParams} 1 {pop} 0 {funcParams} bassStream

    levels &= 0xFFFF // left channel

    float value =# levels
    value /= 32768.0 // 0xFFFF

    value *= 200.0
    
    float y = value
    y /= -2.0
    y += 400.0
    
    use_text_commands {state} true
    draw_rect {x} 0.0 {y} y {width} 50.0 {height} value {rgba} 128 255 128 255
end

terminate_this_custom_script
```

Reply:
try to call `CLEO_GetInternalAudioStream(_, stream)` and compare its result with `bassStream`. I think you're also missing reading the value from the pointer you got. You should return the stream value, not the address

Reply:
This script works in CLEO4. For some reason in CLEO5 it requires
```
int bassStream = stream + 8
```
to work.
I think it should be de referenced once, as this is Bass stream handle value.

Reply:
did you remove the extra dummy class?

Reply:
I added compilation asserts.
Use of dummy for some reason makes it work, otherwise offset is 8.
Static analysis in VS shows offset 4 in both cases.

Reply:
hmm we need to figure out why it happens, otherwise there no guarantee this offset won't become 4 or 12 at some point

Reply:
how can I trust this macro?

![image](https://github.com/user-attachments/assets/9ab502fd-5b25-41f2-af38-f4a0687b2e85)

Reply:
Macro works fine. Correct value is basing on CLEO4 is 4, we have 8 for some reason.
No clue why it happens.

Reply:
I found this question https://stackoverflow.com/questions/27764464/why-is-vtable-pointer-aligned-to-8-bytes-on-x86-compiled-program

It links an article that seemed to have explained the compiler behavior in detail, but unfortunately it's unavailable and internet archive is not working at the moment. But there is an interesting bit:

> The 8 byte structure member causes the vfptr to be misaligned to 8 bytes rather than 4 bytes, as vfptr is added to structure after all the other members have been added/aligned. This behavior seems to be Visual Studio specific

It looks like addition of new `double` properties in CAudioStream changed default alignment from 4 to 8:

With `double`
![image](https://github.com/user-attachments/assets/379de34c-ea24-4689-afec-57a7d9cef347)

With `float`
![image](https://github.com/user-attachments/assets/77328adf-063c-4087-a441-1e7cc17d7c26)

Even though the diagram shows `streamInternal` at offset +4 in both cases, it seems that compiler does some extra optimization and allocates it at offset +8 in the first case. 

Inheriting from an empty virtual class seems to actually help here and avoid this optimization, resulting in correct +4 offset. This trick should also be explained in that article. Here is a relevant discussion https://www.reddit.com/r/programming/comments/11ufrs/the_stolen_bytes_visual_studio_virtual_methods/


With that said, I think we can try the following:

1) make new properties (speed, volume, etc) `float` to enforce 4-byte alignment. I don't see why we need extra precision here.

2) If it's impossible (why?), then we can keep the dummy class.

Thoughts?

Reply:
when IA is back online, we can try find [that article](http://lolengine.net/blog/2012/10/21/the-stolen-bytes) and read the explanation

Reply:
and here is a reproduction using online c++ compiler 
https://cpp.godbolt.org/z/6d8qfqKfW

Reply:
Yes, IA is currently under attack. How did you get these graphs of memory layout?
I changed types to float and it worked.

Reply:
https://devblogs.microsoft.com/visualstudio/size-alignment-and-memory-layout-insights-for-c-classes-structs-and-unions/#memory-layout

Reply:
@MiranDMC IA is back online and the article is here
https://web.archive.org/web/20190115013504/http://lolengine.net/blog/2012/10/21/the-stolen-bytes

TLDR;
> The problem is that the compiler decides to add the vtable pointer after it has aligned the class data, resulting in excessive realignment.